modules/core/src/main/java/org/apache/ignite/internal/processors/diagnostic/DiagnosticProcessor.java

```diff
@@ -53,164 +56,127 @@
+                    if (ctx.config().getDataStorageConfiguration().getPageSize() != DFLT_PAGE_SIZE)
+                        args += " pageSize=" + ctx.config().getDataStorageConfiguration().getPageSize();
 
-                sb.a("]");
+                    args += " pages=" + corruptedPagesFile.getAbsolutePath();
 
-                ignite.log().error(
-                    "Failed to dump diagnostic info on tree corruption. PageIds=" + sb, e);
+                    log.warning(corruptedTreeE.getClass().getSimpleName() + " has occurred. " +
+                        "To diagnose it, make a backup of the following directories: " + walDirsStr + ". " +
+                        "Then, run the following command: java -cp <classpath> " +
+                        "org.apache.ignite.development.utils.IgniteWalConverter " + args);
+                }
+                catch (Throwable t) {
+                    String pages = Arrays.stream(corruptedTreeE.pages())
+                        .map(t2 -> "(" + t2.get1() + ',' + t2.get2() + ')').collect(joining("", "[", "]"));
+
+                    log.error("Failed to dump diagnostic info on tree corruption. PageIds=" + pages, t);
+                }
             }
         }
     }
 
     /**
-     * Resolve file to store diagnostic info.
+     * Creation and filling of a file with pages that can be corrupted.
+     * Pages are written on each line in format "grpId:pageId".
+     * File name format "corruptedPages_TIMESTAMP.txt".
      *
-     * @param customFile Custom file if customized.
-     * @param writeMode Diagnostic file write mode.
-     * @return File to store diagnostic info.
+     * @param dirPath Path to the directory where the file will be created.
+     * @param pages Pages that could be corrupted. Mapping: cache group id -> page id.
+     * @return Created and filled file.
+     * @throws IOException If an I/O error occurs.
      */
-    private File diagnosticFile(File customFile, DiagnosticFileWriteMode writeMode) {
-        if (customFile == null)
-            return finalizeFile(diagnosticPath, writeMode);
+    public static File corruptedPagesFile(Path dirPath, T2<Integer, Long>... pages) throws IOException {
+        dirPath.toFile().mkdirs();
 
-        if (customFile.isAbsolute())
-            return finalizeFile(customFile.toPath(), writeMode);
+        File f = dirPath.resolve("corruptedPages_" + LocalDateTime.now().format(TIME_FORMATTER) + ".txt").toFile();
 
-        return finalizeFile(diagnosticPath.resolve(customFile.toPath()), writeMode);
-    }
+        assert !f.exists();
 
-    /**
-     * @param diagnosticPath Path to diagnostic file.
-     * @param writeMode Diagnostic file write mode.
-     * @return File to store diagnostic info.
-     */
-    private static File finalizeFile(Path diagnosticPath, DiagnosticFileWriteMode writeMode) {
-        diagnosticPath.toFile().mkdirs();
+        try (BufferedWriter bw = new BufferedWriter(new FileWriter(f))) {
```

Comment:
I would suggest to use our FileIO interfaces to get more control over forcing fsync of diagnostics file.